File: packages/protocol/src/tree.ts
import type { SemanticAction, SemanticRole } from './roles.js';
import type { ProvenanceSource } from './probe/ir.js';
import type { CoordinateSpace, Observation, SemanticValueObservation } from './observation.js';
import type { EvidenceProvenance } from './contract.js';

/** Zero-based viewport cell coordinates. */
export interface Rect {
  readonly row: number;
  readonly column: number;
  readonly width: number;
  readonly height: number;
}

/** Closed state set. No arbitrary records. */
export interface SemanticState {
  readonly disabled?: boolean;
  readonly focused?: boolean;
  readonly selected?: boolean;
  readonly checked?: boolean | 'mixed';
  readonly expanded?: boolean;
  readonly modal?: boolean;
  readonly busy?: boolean;
  readonly hidden?: boolean;
  /**
   * The node exists in the layout, but every one of its cells falls outside the
   * visible area — it is scrolled out, and scrolling can bring it back.
   *
   * Named for the claim a test author makes ("this row is off screen"), not for
   * the mechanism that produced it. Clipping is how it happens; being off
   * screen is what it means.
   *
   * **Absent means "not claiming"**, not "on screen". A producer that cannot
   * observe clipping simply omits it. Evidence-qualified geometry remains the
   * authoritative source for viewport and visibility assertions.
   *
   * Implies {@link SemanticState.hidden}: if every cell is outside the visible
   * area then the node is not visible, and validation refuses the pair
   * `offscreen: true` without `hidden: true`.
   */
  readonly offscreen?: boolean;
  readonly readonly?: boolean;
  readonly multiline?: boolean;
  readonly required?: boolean;
  readonly multiselectable?: boolean;
  readonly orientation?: 'horizontal' | 'vertical';
  readonly level?: number;
  readonly positionInSet?: number;
  readonly setSize?: number;
}

/** Authoritative application viewport state in production-defined logical units. */
export interface SemanticScrollState {
  readonly axis: 'vertical' | 'horizontal';
  /** Logical distance from the start of the scrollable content. */
  readonly offset: number;
  /** Logical size of the visible application viewport. */
  readonly viewport: number;
  /** Logical size of the complete scrollable content. */
  readonly extent: number;
}

/** Exact viewport cells painted by one semantic recipient for a committed frame. */
export interface SemanticPaintedRegion {
  /** Bounding box of the painted cells; this is not intended layout geometry. */
  readonly regionBounds: Rect;
  /** Canonical, non-overlapping half-open row runs. */
  readonly spans: readonly { readonly row: number; readonly from: number; readonly to: number }[];
}

/** Maps grapheme offsets of a node's text to cell coordinates (optional capability). */
export interface SemanticTextRange {
  readonly startOffset: number;
  readonly endOffset: number;
  readonly rect: Rect;
}

/**
 * Deterministic JSON data owned by the application domain, not by the portable
 * semantic vocabulary. Containers are allowed, but runtime objects/functions
 * are not; validation applies the protocol's normal depth, byte and collection
 * ceilings recursively.
 */
export interface SemanticExtendedArray extends ReadonlyArray<SemanticExtendedValue> {}

export interface SemanticExtendedObject {
  readonly [key: string]: SemanticExtendedValue;
}

export type SemanticExtendedValue =
  null | boolean | number | string | SemanticExtendedArray | SemanticExtendedObject;

/** Application-defined state, deliberately separate from {@link SemanticState}. */
export type SemanticExtendedState = SemanticExtendedObject;

export interface SemanticNode {
  readonly id: string;
  readonly parentId?: string;
  readonly role: SemanticRole;
  readonly name: string;
  readonly description?: string;
  readonly value?: SemanticValueObservation;
  readonly state?: SemanticState;
  /** Application-specific, serializable state; never promoted to portable flags. */
  readonly extended?: SemanticExtendedState;
  readonly actions?: readonly SemanticAction[];
  /** Authoritative recipe executed only through Termwright's real input devices. */
  readonly inputRecipes?: readonly import('./roles.js').PhysicalInputRecipe[];
  readonly labelledBy?: readonly string[];
  readonly describedBy?: readonly string[];
  readonly textRanges?: readonly SemanticTextRange[];
  /** Author-supplied test id (getByTestId). */
  readonly testId?: string;
  /**
   * The framework's own name for this widget — a class name, a constructor
   * name, a widget type.
   *
   * **Required when `role` is `generic`.** An unrecognised widget must survive
   * as a generic node keeping its geometry, text and children, instead of being
   * dropped with its children reparented. `frameworkType` is what makes such a
   * node identifiable: without it a generic node says only "something was
   * here", which is barely better than the drop it replaced.
   */
  readonly frameworkType?: string;
  /** True when this node may own children the framework probe cannot enumerate. */
  readonly opaqueChildren?: boolean;
  /**
   * Provenance: where this node's facts came from.
   *
   * One source for the whole node, because node facts overwhelmingly share
   * one. Exceptions go in {@link SemanticNode.px}, so a mixed node pays only
   * for the fields that actually differ. Descriptive per-property strings were
   * ruled out by arithmetic — they cost about +91 % against a budget that is
   * already tight.
   */
  readonly p?: ProvenanceSource;
  /** Per-field provenance, for fields whose source differs from `p`. */
  readonly px?: Readonly<Record<string, ProvenanceSource>>;
  /** Evidence-qualified layout facts for this committed observation. */
  readonly geometry: NodeGeometryObservations;
  /** Application scroll state; distinct from the terminal emulator's scrollback. */
  readonly scroll?: Observation<SemanticScrollState>;
  /** Authoritative paint provenance; distinct from layout, clipping and pointer routing. */
  readonly paintedRegion?: Observation<SemanticPaintedRegion>;
}

/** Layout facts reported independently so absence never masquerades as false. */
export interface NodeGeometryObservations {
  readonly displayed: Observation<boolean>;
  readonly intendedRect: Observation<Rect>;
  readonly visibleRect: Observation<Rect>;
}

/** One half-open run of cells with an exact pointer recipient. */
export interface PointerHitRegion {
  /** Canonical non-empty row run: `height` is always 1. */
  readonly rect: Rect;
  readonly recipientId: string;
}

/** A complete point-ownership map for a committed frame. */
export interface PointerHitGrid {
  readonly regions: readonly PointerHitRegion[];
}

/** One provider-owned target region for a committed application revision. */
export interface ProviderPointerRegion {
  readonly recipientId: string;
  /** Bounding box of the pointer-receiving cells; not layout geometry. */
  readonly regionBounds: Rect;
  /** Canonical non-overlapping visible cell runs owned by the target. */
  readonly spans: readonly { readonly row: number; readonly from: number; readonly to: number }[];
}

/** Revision-bound application strategy recipes for one semantic recipient. */
export interface ProviderActionRecipes {
  readonly recipientId: string;
  readonly recipes: readonly import('./roles.js').PhysicalInputRecipe[];
}

/** Revision-bound application viewport state for one semantic recipient. */
export interface ProviderScrollState extends SemanticScrollState {
  readonly recipientId: string;
}

/** Revision-bound application paint provenance for one semantic recipient. */
export interface ProviderPaintedRegion extends SemanticPaintedRegion {
  readonly recipientId: string;
}

/** Production parser configuration hidden by the terminal transport. */
export interface ProviderTerminalInputModes {
  readonly mouseTracking: 'none' | 'x10' | 'vt200' | 'drag' | 'any';
  readonly mouseEncoding: 'default' | 'sgr' | 'urxvt' | 'utf8';
  readonly focusReporting: 'on' | 'off';
}

/** Exact production focus-manager result for one committed revision. */
export type ProviderFocusState =
  { readonly status: 'focused'; readonly recipientId: string } | { readonly status: 'none' };

/**
 * Application evidence bound to the same revision as its containing snapshot.
 * A provider announced in hello must contribute exactly one entry per frame.
 */
export type ProviderRevisionEvidence =
  | {
      readonly providerId: string;
      readonly sessionId: string;
      readonly revision: number;
      readonly status: 'available';
      readonly evidence: EvidenceProvenance;
      readonly pointerRegions: readonly ProviderPointerRegion[];
      /** Exact production focus-manager result when `focus-state` was announced. */
      readonly focusState?: ProviderFocusState;
      /** Production keybinding recipes when `action-recipes` was announced. */
      readonly actionRecipes?: readonly ProviderActionRecipes[];
      /** Complete application viewport facts when `scroll-state` was announced. */
      readonly scrollStates?: readonly ProviderScrollState[];
      /** Complete paint attribution when `painted-regions` was announced. */
      readonly paintedRegions?: readonly ProviderPaintedRegion[];
      /** Production parser configuration when `terminal-input-modes` was announced. */
      readonly inputModes?: ProviderTerminalInputModes;
      /** Complete production-router ownership map when `hit-test` was announced. */
      readonly hitGrid?: PointerHitGrid;
    }
  | {
      readonly providerId: string;
      readonly sessionId: string;
      readonly revision: number;
      readonly status: 'lost';
      readonly reason: string;
    }
  | {
      readonly providerId: string;
      readonly sessionId: string;
      readonly revision: number;
      readonly status: 'violation';
      readonly reason: string;
    };

export interface CursorInfo {
  readonly row: number;
  readonly column: number;
  readonly visible: boolean;
  readonly shape?: 'block' | 'underline' | 'bar';
}

export interface SemanticSnapshot {
  readonly v: 2;
  readonly sessionId: string;
  /** Positive, strictly increasing within a semantic session. */
  readonly revision: number;
  readonly columns: number;
  readonly rows: number;
  readonly cursor?: CursorInfo;
  readonly rootIds: readonly string[];
  readonly nodes: readonly SemanticNode[];
  /** Coordinate system used by every physical observation in this snapshot. */
  readonly coordinateSpace: Observation<CoordinateSpace>;
  /**
   * Required by v2. `known` means a complete map, not a sample or paint-order
   * approximation. Cells absent from a known map have no semantic recipient.
   */
  readonly hitGrid: Observation<PointerHitGrid>;
  /** Present only when application evidence providers were negotiated in hello. */
  readonly providerEvidence?: readonly ProviderRevisionEvidence[];
}

File: clients/go/protocol/tree.go
package protocol

// Rect is a zero-based viewport cell rectangle.
type Rect struct {
	Row    int `json:"row"`
	Column int `json:"column"`
	Width  int `json:"width"`
	Height int `json:"height"`
}

// Observation preserves whether a fact is known, absent, retryably unknown,
// or unsupported. Value is meaningful only when Status is "known".
type Observation[T any] struct {
	Status     string              `json:"status"`
	Value      *T                  `json:"value,omitempty"`
	Evidence   *EvidenceProvenance `json:"evidence,omitempty"`
	Reason     string              `json:"reason,omitempty"`
	Capability string              `json:"capability,omitempty"`
}

// EvidenceProvenance identifies how a known or authoritatively absent
// observation was established.
type EvidenceProvenance struct {
	Source     string `json:"source"`
	Method     string `json:"method"`
	Strength   string `json:"strength"`
	ProviderID string `json:"providerId"`
}

// SemanticValueObservation preserves value support, absence and confidentiality.
// Value is present only for Status "known"; a withheld sensitive value never
// crosses the protocol boundary.
type SemanticValueObservation struct {
	Status      string              `json:"status"`
	Value       *string             `json:"value,omitempty"`
	Sensitivity string              `json:"sensitivity,omitempty"`
	Evidence    *EvidenceProvenance `json:"evidence,omitempty"`
	Reason      string              `json:"reason,omitempty"`
	Capability  string              `json:"capability,omitempty"`
}

type NodeGeometryObservations struct {
	Displayed    Observation[bool] `json:"displayed"`
	IntendedRect Observation[Rect] `json:"intendedRect"`
	VisibleRect  Observation[Rect] `json:"visibleRect"`
}

type PointerHitRegion struct {
	Rect        Rect   `json:"rect"`
	RecipientID string `json:"recipientId"`
}

type PointerHitGrid struct {
	Regions []PointerHitRegion `json:"regions"`
}

// ProviderPointerSpan is one half-open row run owned by a recipient.
type ProviderPointerSpan struct {
	Row  int `json:"row"`
	From int `json:"from"`
	To   int `json:"to"`
}

// ProviderPointerRegion is pointer-only evidence. RegionBounds is not layout
// geometry or visual clipping.
type ProviderPointerRegion struct {
	RecipientID  string                `json:"recipientId"`
	RegionBounds Rect                  `json:"regionBounds"`
	Spans        []ProviderPointerSpan `json:"spans"`
}

// PaintedRegion is authoritative cell attribution from a production painter.
type PaintedRegion struct {
	RegionBounds Rect                  `json:"regionBounds"`
	Spans        []ProviderPointerSpan `json:"spans"`
}

// ProviderPaintedRegion binds painted cells to one semantic recipient.
type ProviderPaintedRegion struct {
	RecipientID  string                `json:"recipientId"`
	RegionBounds Rect                  `json:"regionBounds"`
	Spans        []ProviderPointerSpan `json:"spans"`
}

// ProviderActionRecipes binds production keybindings to one semantic recipient.
type ProviderActionRecipes struct {
	RecipientID string                `json:"recipientId"`
	Recipes     []PhysicalInputRecipe `json:"recipes"`
}

// ProviderFocusState is the exact production focus-manager result.
type ProviderFocusState struct {
	Status      string `json:"status"`
	RecipientID string `json:"recipientId,omitempty"`
}

// ProviderScrollState is application viewport state in logical units.
type ProviderScrollState struct {
	RecipientID string `json:"recipientId"`
	Axis        string `json:"axis"`
	Offset      int    `json:"offset"`
	Viewport    int    `json:"viewport"`
	Extent      int    `json:"extent"`
}

// ProviderTerminalInputModes is the application's production terminal parser
// configuration for one committed semantic revision.
type ProviderTerminalInputModes struct {
	MouseTracking  string `json:"mouseTracking"`
	MouseEncoding  string `json:"mouseEncoding"`
	FocusReporting string `json:"focusReporting"`
}

// ProviderRevisionEvidence binds one frozen application provider to a session
// and semantic revision. Status is available, lost, or violation.
type ProviderRevisionEvidence struct {
	ProviderID     string                      `json:"providerId"`
	SessionID      string                      `json:"sessionId"`
	Revision       int64                       `json:"revision"`
	Status         string                      `json:"status"`
	Evidence       *EvidenceProvenance         `json:"evidence,omitempty"`
	PointerRegions *[]ProviderPointerRegion    `json:"pointerRegions,omitempty"`
	FocusState     *ProviderFocusState         `json:"focusState,omitempty"`
	ActionRecipes  *[]ProviderActionRecipes    `json:"actionRecipes,omitempty"`
	ScrollStates   *[]ProviderScrollState      `json:"scrollStates,omitempty"`
	PaintedRegions *[]ProviderPaintedRegion    `json:"paintedRegions,omitempty"`
	InputModes     *ProviderTerminalInputModes `json:"inputModes,omitempty"`
	HitGrid        *PointerHitGrid             `json:"hitGrid,omitempty"`
	Reason         string                      `json:"reason,omitempty"`
}

// State is the closed state set. Pointers distinguish "unset" from "false":
// an omitted state is not an assertion, a false one is.
type State struct {
	Disabled *bool `json:"disabled,omitempty"`
	Focused  *bool `json:"focused,omitempty"`
	Selected *bool `json:"selected,omitempty"`
	Checked  any   `json:"checked,omitempty"` // bool or the string "mixed"
	Expanded *bool `json:"expanded,omitempty"`
	Modal    *bool `json:"modal,omitempty"`
	Busy     *bool `json:"busy,omitempty"`
	Hidden   *bool `json:"hidden,omitempty"`
	// Offscreen says every cell is outside the visible area — scrolled away,
	// not undisplayed. Implies Hidden; the pair without it is refused.
	Offscreen       *bool  `json:"offscreen,omitempty"`
	ReadOnly        *bool  `json:"readonly,omitempty"`
	Multiline       *bool  `json:"multiline,omitempty"`
	Required        *bool  `json:"required,omitempty"`
	Multiselectable *bool  `json:"multiselectable,omitempty"`
	Orientation     string `json:"orientation,omitempty"`
	Level           *int   `json:"level,omitempty"`
	PositionInSet   *int   `json:"positionInSet,omitempty"`
	SetSize         *int   `json:"setSize,omitempty"`
}

// ScrollState is application viewport state, distinct from terminal scrollback.
type ScrollState struct {
	Axis     string `json:"axis"`
	Offset   int    `json:"offset"`
	Viewport int    `json:"viewport"`
	Extent   int    `json:"extent"`
}

// TextRange maps grapheme offsets of a node's text onto cell coordinates.
type TextRange struct {
	StartOffset int  `json:"startOffset"`
	EndOffset   int  `json:"endOffset"`
	Rect        Rect `json:"rect"`
}

// PhysicalInputRecipeStep is data executed later by Termwright's PTY devices.
type PhysicalInputRecipeStep struct {
	Kind string `json:"kind"`
	Key  string `json:"key,omitempty"`
}

type PhysicalInputRecipe struct {
	Action        string                    `json:"action"`
	RequiresFocus bool                      `json:"requiresFocus"`
	Steps         []PhysicalInputRecipeStep `json:"steps"`
}

// Node is one accessible element.
type Node struct {
	ID          string `json:"id"`
	ParentID    string `json:"parentId,omitempty"`
	Role        Role   `json:"role"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	// Value is what the widget contains, with absence and confidentiality kept
	// distinct from an empty public string.
	Value *SemanticValueObservation `json:"value,omitempty"`
	State *State                    `json:"state,omitempty"`
	// Extended is application-defined JSON state. It is deliberately separate
	// from State, whose portable vocabulary remains closed.
	Extended     map[string]any        `json:"extended,omitempty"`
	Actions      []Action              `json:"actions,omitempty"`
	InputRecipes []PhysicalInputRecipe `json:"inputRecipes,omitempty"`
	LabelledBy   []string              `json:"labelledBy,omitempty"`
	DescribedBy  []string              `json:"describedBy,omitempty"`
	TextRanges   []TextRange           `json:"textRanges,omitempty"`
	TestID       string                `json:"testId,omitempty"`
	// FrameworkType is what the UI framework calls this widget. Required when
	// Role is RoleGeneric: an unrecognised widget must at least name its own
	// type, so a reader can tell one unknown thing from another.
	FrameworkType string `json:"frameworkType,omitempty"`
	// P is where this node's facts came from, as a whole.
	P string `json:"p,omitempty"`
	// PX is where individual fields came from, when they differ from P.
	PX            map[string]string           `json:"px,omitempty"`
	Geometry      NodeGeometryObservations    `json:"geometry"`
	Scroll        *Observation[ScrollState]   `json:"scroll,omitempty"`
	PaintedRegion *Observation[PaintedRegion] `json:"paintedRegion,omitempty"`
}

// Provenance sources: where a semantic fact came from. Closed set.
const (
	ProvenanceAnnotation  = "annotation"
	ProvenanceRecognizer  = "recognizer"
	ProvenanceFramework   = "framework"
	ProvenanceApplication = "application"
	ProvenanceCorrelation = "correlation"
	ProvenanceHeuristic   = "heuristic"
)

// ProvenanceSources is every value P and PX accept.
var ProvenanceSources = []string{
	ProvenanceAnnotation, ProvenanceRecognizer, ProvenanceFramework, ProvenanceApplication,
	ProvenanceCorrelation, ProvenanceHeuristic,
}

// Cursor is the terminal cursor position in viewport cells.
type Cursor struct {
	Row     int    `json:"row"`
	Column  int    `json:"column"`
	Visible bool   `json:"visible"`
	Shape   string `json:"shape,omitempty"`
}

// Snapshot is the whole tree for one committed render.
type Snapshot struct {
	V                int                         `json:"v"`
	SessionID        string                      `json:"sessionId"`
	Revision         int64                       `json:"revision"`
	Columns          int                         `json:"columns"`
	Rows             int                         `json:"rows"`
	Cursor           *Cursor                     `json:"cursor,omitempty"`
	RootIDs          []string                    `json:"rootIds"`
	Nodes            []Node                      `json:"nodes"`
	CoordinateSpace  Observation[string]         `json:"coordinateSpace"`
	HitGrid          Observation[PointerHitGrid] `json:"hitGrid"`
	ProviderEvidence []ProviderRevisionEvidence  `json:"providerEvidence,omitempty"`
}

// NewSnapshot returns an evidence-qualified snapshot. Callers must populate
// every node's Geometry and explicitly qualify all observable facts.
func NewSnapshot(sessionID string, revision int64, columns, rows int) *Snapshot {
	space := "viewport-cells"
	return &Snapshot{
		V: 2, SessionID: sessionID, Revision: revision, Columns: columns, Rows: rows,
		RootIDs: []string{}, Nodes: []Node{},
		CoordinateSpace: Observation[string]{Status: "known", Value: &space, Evidence: DefaultEvidence("semantic-adapter")},
		HitGrid:         Observation[PointerHitGrid]{Status: "unsupported", Capability: string(CapPointerHitGrid), Reason: "not-negotiated"},
	}
}

// DefaultEvidence describes an authoritative fact declared by an adapter.
func DefaultEvidence(providerID string) *EvidenceProvenance {
	return &EvidenceProvenance{Source: "application", Method: "declared", Strength: "authoritative", ProviderID: providerID}
}

// Bool returns a pointer to v, for the optional State fields.
func Bool(v bool) *bool { return &v }

// Int returns a pointer to v, for the optional State fields.
func Int(v int) *int { return &v }

// PublicValue creates an authoritative public semantic value.
func PublicValue(v string, evidence *EvidenceProvenance) *SemanticValueObservation {
	return &SemanticValueObservation{Status: "known", Value: &v, Sensitivity: "public", Evidence: evidence}
}

// WithheldSensitiveValue declares that a value exists but must not cross the wire.
func WithheldSensitiveValue() *SemanticValueObservation {
	return &SemanticValueObservation{Status: "withheld", Reason: "sensitive", Sensitivity: "sensitive"}
}
